pkg/standalone-utils/contracts/PriceImpactHelper.sol

```diff
@@ -0,0 +1,251 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import { IRouter } from "@balancer-labs/v3-interfaces/contracts/vault/IRouter.sol";
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+
+import { CallAndRevert } from "./CallAndRevert.sol";
+
+contract PriceImpactHelper is CallAndRevert {
+    using FixedPoint for uint256;
+
+    IVault internal immutable _vault;
+    IRouter internal immutable _router;
+
+    constructor(IVault vault, IRouter router) {
+        _vault = vault;
+        _router = router;
+    }
+
+    /*******************************************************************************
+                                Price Impact
+    *******************************************************************************/
+
+    function calculateAddLiquidityUnbalancedPriceImpact(
+        address pool,
+        uint256[] memory exactAmountsIn,
+        address sender
+    ) external returns (uint256 priceImpact) {
+        uint256 bptAmountOut = _queryAddLiquidityUnbalanced(pool, exactAmountsIn, sender);
+        uint256[] memory proportionalAmountsOut = _queryRemoveLiquidityProportional(pool, bptAmountOut, sender);
+
+        // get deltas between exactAmountsIn and proportionalAmountsOut
+        int256[] memory deltas = new int256[](exactAmountsIn.length);
+        for (uint256 i = 0; i < exactAmountsIn.length; i++) {
+            deltas[i] = int(proportionalAmountsOut[i]) - int(exactAmountsIn[i]);
+        }
+
+        // query add liquidity for each delta, so we know how unbalanced each amount in is in terms of BPT
+        int256[] memory deltaBPTs = new int256[](exactAmountsIn.length);
+        for (uint256 i = 0; i < exactAmountsIn.length; i++) {
+            deltaBPTs[i] = _queryAddLiquidityUnbalancedForTokenDeltas(pool, i, deltas, sender);
+        }
+
+        // zero out deltas leaving only a remaining delta within a single token
+        uint256 remainingDeltaIndex = _zeroOutDeltas(pool, deltas, deltaBPTs, sender);
+
+        // calculate price impact ABA with remaining delta and its respective exactAmountIn
+        // remaining delta is always negative, so by multiplying by -1 we get a positive number
+        uint256 delta = uint(-deltas[remainingDeltaIndex]);
```

Comment:
I'd still just use safe cast here. Everything is static so we don't care about gas, and we have plenty of bytecode to spare.